crates/types/src/simple_certificate.rs

```diff
@@ -0,0 +1,129 @@
+    pub leaf_commitment: VOTEABLE,
+    /// commitment of all the votes this cert should be signed over
+    pub vote_commitment: Commitment<VOTEABLE>,
+    /// Which view this QC relates to
+    pub view_number: TYPES::Time,
+    /// assembled signature for certificate aggregation
+    pub signatures: Option<<TYPES::SignatureKey as SignatureKey>::QCType>,
+    /// If this QC is for the genesis block
+    pub is_genesis: bool,
+    /// phantom data for `MEMBERSHIP` and `TYPES`
+    pub _pd: PhantomData<TYPES>,
+}
+
+impl<TYPES: NodeType, VOTEABLE: Voteable + 'static> Certificate2<TYPES>
+    for SimpleCertificate<TYPES, VOTEABLE>
+{
+    type Voteable = VOTEABLE;
+    // type Membership = MEMBERSHIP;
+
+    fn create_signed_certificate(
+        vote_commitment: Commitment<VOTEABLE>,
+        data: Self::Voteable,
+        sig: <TYPES::SignatureKey as SignatureKey>::QCType,
+        view: TYPES::Time,
+    ) -> Self {
+        SimpleCertificate {
+            leaf_commitment: data,
+            vote_commitment,
+            view_number: view,
+            signatures: Some(sig),
+            is_genesis: false,
+            _pd: PhantomData,
+        }
+    }
+    fn is_valid_cert<MEMBERSHIP: Membership<TYPES>>(&self, membership: &MEMBERSHIP) -> bool {
+        if self.is_genesis && self.view_number == TYPES::Time::genesis() {
+            return true;
+        }
+        let real_qc_pp = <TYPES::SignatureKey as SignatureKey>::get_public_parameter(
+            membership.get_committee_qc_stake_table(),
+            U256::from(membership.success_threshold().get()),
+        );
+        <TYPES::SignatureKey as SignatureKey>::check(
+            &real_qc_pp,
+            self.vote_commitment.as_ref(),
+            self.signatures.as_ref().unwrap(),
+        )
+    }
+    fn threshold<MEMBERSHIP: Membership<TYPES>>(membership: &MEMBERSHIP) -> u64 {
+        membership.success_threshold().into()
+    }
+    fn get_data(&self) -> &Self::Voteable {
+        &self.leaf_commitment
+    }
+    fn get_data_commitment(&self) -> Commitment<Self::Voteable> {
+        self.vote_commitment
+    }
+}
+
+impl<TYPES: NodeType, VOTEABLE: Voteable + 'static> HasViewNumber<TYPES>
+    for SimpleCertificate<TYPES, VOTEABLE>
+{
+    fn get_view_number(&self) -> TYPES::Time {
+        self.view_number
+    }
+}
+impl<TYPES: NodeType, VOTEABLE: Voteable + 'static> Display
+    for QuorumCertificate2<TYPES, VOTEABLE>
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "view: {:?}, is_genesis: {:?}",
+            self.view_number, self.is_genesis
+        )
+    }
+}
+
+impl<
+        TYPES: NodeType,
+        LEAF: Committable + Committable + Clone + Serialize + Debug + PartialEq + Hash + Eq + 'static,
+    > SimpleCertificate<TYPES, YesData<LEAF>>
+{
+    #[must_use]
+    /// Creat the Genisis certificate
+    pub fn genesis() -> Self {
```

Comment:
This is only used for the QuorumCertificate.  It shouldn't ever be used on other certificates.  Because of that I feel like we shouldn't have it here.  Ideally we'd want to protect from accidentally calling it on the other certificate types.

Reply:
I believe the way this is implemented is that it's only for `SimpleCertificate`s with VOTABLE type `YesData` which means any other certificate types calling genesis it will be unimplemented.  I'll use the `QurorumCertificate2` type alias here instead to make it more clear but it's functionally the same.

Reply:
Oh I see, yes you are right! I didn't look closely enough at the type signature.